src/Database/LSMTree/Internal/Run.hs

```diff
@@ -30,4 +32,115 @@
+
+       -- | The in-memory index mapping keys to page numbers in the
+       -- Key\/Ops file. In future we may support alternative index
+       -- representations.
+       lsmRunIndex    :: !CompactIndex,
+
+       -- | The file handle for the Key\/Ops file. This file is opened
+       -- read-only and is accessed in a page-oriented way, i.e. only
+       -- reading whole pages, at page offsets. It will be opened with
+       -- 'O_DIRECT' on supported platforms.
+       lsmRunKOpsFile :: !fhandle,
+
+       -- | The file handle for the BLOBs file. This file is opened
+       -- read-only and is accessed in a normal style using buffered
+       -- I\/O, reading arbitrary file offset and length spans.
+       lsmRunBlobFile :: !fhandle
+     }
+
+addReference :: HasFS IO h -> Run (FS.Handle h) -> IO ()
+addReference _ Run {..} =
+    atomicModifyIORef lsmRunRefCount (\n -> (n+1, ()))
+
+-- | After calling this operation, the run must not be used anymore.
+-- Might cause the run to be closed, which removes all its associated
+-- files from disk.
+removeReference :: HasFS IO h -> Run (FS.Handle h) -> IO ()
+removeReference fs run@Run {..} = do
+    count <- atomicModifyIORef' lsmRunRefCount (\n -> (n-1, n-1))
+    when (count <= 0) $
+      close fs run
+
+-- | After calling this operation, the run must not be used anymore.
+-- Removes all files associated with the run from disk.
+close :: HasFS IO h -> Run (FS.Handle h) -> IO ()
+close fs Run {..} = do
+    -- close and remove all files
+    -- TODO: even if one of them throws an exception
+    FS.hClose fs lsmRunKOpsFile
+    FS.hClose fs lsmRunBlobFile
+    FS.removeFile fs (runKOpsPath lsmRunFsPaths)
+    FS.removeFile fs (runBlobPath lsmRunFsPaths)
+    FS.removeFile fs (runFilterPath lsmRunFsPaths)
+    FS.removeFile fs (runIndexPath lsmRunFsPaths)
+
+fromMutable :: HasFS IO h -> MRun (FS.Handle h) -> IO (Run (FS.Handle h))
+fromMutable fs mrun = do
+    (lsmRunRefCount, lsmRunFsPaths, lsmRunFilter, lsmRunIndex) <-
+      unsafeFinalise fs mrun
+    -- open read mode
+    lsmRunKOpsFile <- FS.hOpen fs (runKOpsPath lsmRunFsPaths) FS.ReadMode
+    lsmRunBlobFile <- FS.hOpen fs (runBlobPath lsmRunFsPaths) FS.ReadMode
+    return Run {..}
+
+
+-- | Writes the entries to disk, including the blobs they contain.
+--
+-- As a possible optimisation, blobs could be written to a blob file immediately
+-- when they are added to the write buffer, avoiding the need to do it here.
+--
+-- To create the compact index, we need to estimate the number of pages.
+-- This is currently done very crudely based on the assumption that a k\/op pair
+-- requires approximately 100 bytes of disk space.
+-- To do better, we would need information about key and value size.
+fromWriteBuffer :: HasFS IO h -> RunFsPaths -> WriteBuffer k v b ->
+                   IO (Run (FS.Handle h))
+fromWriteBuffer fs fsPaths buffer = do
+    mrun <- new fs fsPaths numEntries numPages
+    for_ (WB.content buffer) $ \(k, e) ->
```

Comment:
I'm not sure how much allocation this creates, but on the other hand I'd be worried about `traverseWithKey` returning an unnecessary updated `Map`. We'd want`traverseWithKey_`, I think the best we can get is something like a `foldrWithKey' (*>)`? I can try that, but I don't know if it will perform better.